### PR TITLE
Refactoring code contracts implementation

### PR DESCRIPTION
**Pull request by feliperodri**
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-instrument/contracts/assigns.cpp [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+    {
+      exprt::operandst conjuncts;
+      conjuncts.push_back(result);
+      conjuncts.push_back(current_target_compatible);
+      result = conjunction(conjuncts);
+    }
+  }
+
+  return result;
+}
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/assigns.cpp]: [nit]

Missing newline

File: regression/contracts/assigns_replace_06/test.desc [hunk 2]
```diff
@@ -1,9 +1,9 @@
 CORE
 main.c
 --replace-all-calls-with-contracts
-^EXIT=0$
+^EXIT=10$
 ^SIGNAL=0$
-^VERIFICATION SUCCESSFUL$
+^VERIFICATION FAILED$
```
Comment (SaswatPadhi) [inline regression/contracts/assigns_replace_06/test.desc:6]: This change looks scary in a "refactoring" PR. Why was this not caught before? ...

File: src/goto-instrument/contracts/contracts.h [hunk 3]
```diff
@@ -220,331 +246,4 @@ class code_contractst
...
-  function_binding_visitort() : const_expr_visitort()
-  {
-  }
-
-  void operator()(const exprt &exp) override
-  {
-  }
-};
-
 #endif // CPROVER_GOTO_INSTRUMENT_CODE_CONTRACTS_H
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/contracts.h]: The header guard no longer matches the path.

```suggestion
#endif // CPROVER_GOTO_INSTRUMENT_CONTRACTS_CONTRACTS_H
```

Also update at the top of this file

File: src/goto-instrument/contracts/contracts.cpp [hunk 4]
```diff
@@ -1041,7 +975,7 @@ bool code_contractst::enforce_contract(const std::string &fun_to_enforce)
   auto nexist_old_function = goto_functions.function_map.find(original);
   INVARIANT(
     nexist_old_function == goto_functions.function_map.end(),
-    "There should be no function called " + fun_to_enforce +
+    "There should be no function called " + id2string(function) +
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/contracts.cpp:979]: Should this be `function` (which is used above in several places), or `id2string(function)`?

File: src/goto-instrument/contracts/contracts.cpp [hunk 5]
```diff
@@ -11,98 +11,28 @@ Date: February 2016
 /// \file
 /// Verify and use annotated invariants and pre/post-conditions
 
-#include "code_contracts.h"
+#include "contracts.h"
+#include "assigns.h"
+#include "memory_predicates.h"
```
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/contracts.cpp:17]: [nit]

Consider sorting these

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]
Overall LGTM, some minor comments below:

File: src/goto-instrument/contracts/contracts.h (hunk 3 above)
Comment (feliperodri) [reply src/goto-instrument/contracts/contracts.h]: Fixed it.

File: src/goto-instrument/contracts/contracts.cpp (hunk 5 above)
Comment (feliperodri) [reply src/goto-instrument/contracts/contracts.cpp:17]: `clang-format` adds the corresponding `.h` at the top. I added a new line.

File: src/goto-instrument/contracts/contracts.cpp (hunk 4 above)
Comment (feliperodri) [reply src/goto-instrument/contracts/contracts.cpp:979]: It should be `id2string(function)` to use the correct overloaded operator +.

File: regression/contracts/assigns_replace_06/test.desc (hunk 2 above)
Comment (feliperodri) [reply regression/contracts/assigns_replace_06/test.desc:6]: Before, we were not havocing `array[]` correctly and this test description was wrong. Take a look at a snippet from this test case
```c
void foo(char c[]) __CPROVER_assigns(c)
{
}

void bar(char d[]) __CPROVER_assigns(d)
{
}

int main()
{
  char b[10];
  b[0] = 'a';
  b[1] = 'b';
  b[2] = 'c';
  b[3] = 'd';
  b[4] = 'e';
  b[5] = 'f';
  b[6] = 'g';
  b[7] = 'h';
  b[8] = 'i';
  b[9] = 'j';
  foo(b);
  assert(b[0] == 'a');
  /* ... */
```

During replacement, this should clearly fail. We still need to add more checks (on a separate PR), to ensure we are also havoc the contents of arrays correctly (see https://github.com/diffblue/cbmc/issues/6235).

File: src/goto-instrument/contracts/contracts.cpp (hunk 4 above)
Comment (SaswatPadhi) [reply src/goto-instrument/contracts/contracts.cpp:979]: Oh right, thanks!

**Review by SaswatPadhi — APPROVED** [review]
Thanks for the refactoring! LGTM

File: regression/contracts/assigns_replace_06/test.desc (hunk 2 above)
Comment (martin-cs) [reply regression/contracts/assigns_replace_06/test.desc:6]: ( Could this test case make it into one of those PRs please? )

**Review by martin-cs — APPROVED** [review]
This kind of refactor is often hard to review so at best I think this is OK and I trust @feliperodri 's judgement here.  There is a 40% reduction in code or so and it doesn't break the tests so that seems like a good start.

File: regression/contracts/assigns_replace_06/test.desc (hunk 2 above)
Comment (feliperodri) [reply regression/contracts/assigns_replace_06/test.desc:6]: Yes @martin-cs I added a test case to the GitHub issue, so I can make sure it gets into the PR that fixes it.

**Review by chrisr-diffblue — APPROVED** [review]

**Pull request merged**